### PR TITLE
Tu clothing spider

### PR DESCRIPTION
**Pull request by rosheen33**

### CODE REVIEW

File: tu_spider/tu_spider/items.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    brand = scrapy.Field()
+    image_urls = scrapy.Field()
+    skus = scrapy.Field()
+    care = scrapy.Field()
+    name = scrapy.Field()
+    gender = scrapy.Field()
+    merch_info = scrapy.Field()
+
+
+class skuItem(scrapy.Item):
```
Comment (alisufian) [inline tu_spider/tu_spider/items.py]: Incorrect casing

File: tu_spider/tu_spider/spiders/tu_crawler.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        # Make a dictionary
+        skus_collection = {}
+        skus.pop(0)
+        url_domain = response.url
+
+        for s in skus:
+            # Making full URL
+            url_of_sku = urljoin(url_domain, s)
+            data = urllib2.urlopen(url_of_sku).read()
+            skus_collection[(s).split('=')[-1]] = self.parse_skus(data)
```
Comment (alisufian) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: This is not the way to do it. No direct http calls from inside the spider.

Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Guide me toward the right way

On 3 September 2015 at 18:50, Ali Sufian <notifications@github.com> wrote:

> In tu_spider/tu_spider/spiders/tu_crawler.py
> <https://github.com/arbisoft/the-lab/pull/60#discussion_r38647111>:
>
> > +        size = re.sub(ur'\D', u'', size, flags=re.UNICODE) + " years"
> > +        item["size"] = size
> > +        return item
> > +
> > +    def get_skus(self, skus, response):
> > +
> > +        # Make a dictionary
> > +        skus_collection = {}
> > +        skus.pop(0)
> > +        url_domain = response.url
> > +
> > +        for s in skus:
> > +            # Making full URL
> > +            url_of_sku = urljoin(url_domain, s)
> > +            data = urllib2.urlopen(url_of_sku).read()
> > +            skus_collection[(s).split('=')[-1]] = self.parse_skus(data)
>
> This is not the way to do it. No direct http calls from inside the spider.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60/files#r38647111>.
>



-- 

*Rosheen Pervaiz*

Comment (alisufian) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Use scrapy to request.

File: tu_spider/tu_spider/items.py (hunk 1 above)
Comment (rosheen33) [inline tu_spider/tu_spider/items.py]: Corrected

On 3 September 2015 at 18:48, Ali Sufian <notifications@github.com> wrote:

> In tu_spider/tu_spider/items.py
> <https://github.com/arbisoft/the-lab/pull/60#discussion_r38646952>:
>
> > +class TuSpiderItem(scrapy.Item):
> > +    # define the fields for your item here like:
> > +    category = scrapy.Field()
> > +    product_id = scrapy.Field()
> > +    description = scrapy.Field()
> > +    url_orignal = scrapy.Field()
> > +    brand = scrapy.Field()
> > +    image_urls = scrapy.Field()
> > +    skus = scrapy.Field()
> > +    care = scrapy.Field()
> > +    name = scrapy.Field()
> > +    gender = scrapy.Field()
> > +    merch_info = scrapy.Field()
> > +
> > +
> > +class skuItem(scrapy.Item):
>
> Incorrect casing
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60/files#r38646952>.
>



-- 

*Rosheen Pervaiz*

File: tu_spider/tu_spider/spiders/tu_crawler.py (hunk 2 above)
Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: should I use "scrapy.http.Request" ?

On 3 September 2015 at 18:56, Ali Sufian <notifications@github.com> wrote:

> In tu_spider/tu_spider/spiders/tu_crawler.py
> <https://github.com/arbisoft/the-lab/pull/60#discussion_r38647759>:
>
> > +        size = re.sub(ur'\D', u'', size, flags=re.UNICODE) + " years"
> > +        item["size"] = size
> > +        return item
> > +
> > +    def get_skus(self, skus, response):
> > +
> > +        # Make a dictionary
> > +        skus_collection = {}
> > +        skus.pop(0)
> > +        url_domain = response.url
> > +
> > +        for s in skus:
> > +            # Making full URL
> > +            url_of_sku = urljoin(url_domain, s)
> > +            data = urllib2.urlopen(url_of_sku).read()
> > +            skus_collection[(s).split('=')[-1]] = self.parse_skus(data)
>
> Use scrapy to request.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60/files#r38647759>.
>



-- 

*Rosheen Pervaiz*

File: tu_spider/tu_spider/spiders/tu_crawler.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+from scrapy.spiders import CrawlSpider, Rule
+from scrapy.linkextractors.sgml import SgmlLinkExtractor
+from scrapy.linkextractors import LinkExtractor
+import logging
+import requests
+import urllib2
+from urlparse import urljoin
+from scrapy.shell import inspect_response
+
+
```
Comment (alisufian) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Please remove unused imports.

Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: okay

On 7 September 2015 at 14:18, Ali Sufian <notifications@github.com> wrote:

> In tu_spider/tu_spider/spiders/tu_crawler.py
> <https://github.com/arbisoft/the-lab/pull/60#discussion_r38845988>:
>
> > +from scrapy.spiders import BaseSpider
> > +from scrapy.selector import HtmlXPathSelector
> > +from tu_spider.items import TuSpiderItem
> > +from tu_spider.items import SkuItem
> > +from scrapy.http import Request
> > +from scrapy.utils.serialize import ScrapyJSONEncoder
> > +from scrapy.spiders import CrawlSpider, Rule
> > +from scrapy.linkextractors.sgml import SgmlLinkExtractor
> > +from scrapy.linkextractors import LinkExtractor
> > +import logging
> > +import requests
> > +import urllib2
> > +from urlparse import urljoin
> > +from scrapy.shell import inspect_response
> > +
> > +
>
> Please remove unused imports.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60/files#r38845988>.
>



-- 

*Rosheen Pervaiz*

Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Done

**Comment by alisufian** [comment]
Looking at the code it seems each item will contain only one sku. Is this correct ?

File: tu_spider/tu_spider/spiders/tu_crawler.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        for i in self.get_skus(hxs.select("//select[@id='Size']/option/@value").extract(), item, response.url):
+            yield i
+
+    def get_skus(self, skus, item, url):
+
+        # Make a dictionary
+        skus_collection = {}
+        url_domain = url
+        for s in skus[1:]:
```
Comment (alisufian) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: skus_collection is not used anywhere and url_domain is also not required.

**Comment by rosheen33** [comment]
No , each item will get all skus iteratively

On 7 September 2015 at 14:56, Ali Sufian <notifications@github.com> wrote:

> Looking at the code it seems each item will contain only one sku. Is this
> correct ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60#issuecomment-138256258>.
>



-- 

*Rosheen Pervaiz*

File: tu_spider/tu_spider/spiders/tu_crawler.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        care_instructions[1] = care_instructions[1].strip()
+        gender = category[0]
+
+        item["product_id"] = product_id
+        item["care"] = care_instructions
+        item["category"] = category
+        item["name"] = name
+        item["description"] = description
+        item["image_urls"] = image_urls
+        item["url_orignal"] = url_orignal
```
Comment (alisufian) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: incorrect spelling `orignal`

File: tu_spider/tu_spider/spiders/tu_crawler.py (hunk 4 above)
Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Removed

File: tu_spider/tu_spider/spiders/tu_crawler.py (hunk 5 above)
Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Corrected

File: tu_spider/tu_spider/spiders/tu_crawler.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        item["category"] = category
+        item["name"] = name
+        item["description"] = description
+        item["image_urls"] = image_urls
+        item["url_original"] = url_original
+        item["brand"] = brand
+        item["gender"] = gender
+        item['skus'] = {}
+
+        for i in self.get_skus(hxs.select("//select[@id='Size']/option/@value").extract(), item, response.url):
```
Comment (alisufian) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Lets not make such a long call inside the for. Its less readable.

Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: okay.

On 7 September 2015 at 18:01, Ali Sufian <notifications@github.com> wrote:

> In tu_spider/tu_spider/spiders/tu_crawler.py
> <https://github.com/arbisoft/the-lab/pull/60#discussion_r38860252>:
>
> > +            "//div[@class='productDetails_care_instructions']/text()[2]").extract()
> > +        care_instructions[1] = care_instructions[1].strip()
> > +        gender = category[0]
> > +
> > +        item["product_id"] = product_id
> > +        item["care"] = care_instructions
> > +        item["category"] = category
> > +        item["name"] = name
> > +        item["description"] = description
> > +        item["image_urls"] = image_urls
> > +        item["url_original"] = url_original
> > +        item["brand"] = brand
> > +        item["gender"] = gender
> > +        item['skus'] = {}
> > +
> > +        for i in self.get_skus(hxs.select("//select[@id='Size']/option/@value").extract(), item, response.url):
>
> Lets not make such a long call inside the for. Its less readable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60/files#r38860252>.
>



-- 

*Rosheen Pervaiz*

Comment (rosheen33) [inline tu_spider/tu_spider/spiders/tu_crawler.py]: Removed the "For" loop

On 7 September 2015 at 18:01, Ali Sufian <notifications@github.com> wrote:

> In tu_spider/tu_spider/spiders/tu_crawler.py
> <https://github.com/arbisoft/the-lab/pull/60#discussion_r38860252>:
>
> > +            "//div[@class='productDetails_care_instructions']/text()[2]").extract()
> > +        care_instructions[1] = care_instructions[1].strip()
> > +        gender = category[0]
> > +
> > +        item["product_id"] = product_id
> > +        item["care"] = care_instructions
> > +        item["category"] = category
> > +        item["name"] = name
> > +        item["description"] = description
> > +        item["image_urls"] = image_urls
> > +        item["url_original"] = url_original
> > +        item["brand"] = brand
> > +        item["gender"] = gender
> > +        item['skus'] = {}
> > +
> > +        for i in self.get_skus(hxs.select("//select[@id='Size']/option/@value").extract(), item, response.url):
>
> Lets not make such a long call inside the for. Its less readable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/arbisoft/the-lab/pull/60/files#r38860252>.
>



-- 

*Rosheen Pervaiz*

**Pull request merged**